### DISCUSSION TITLE
Clarification on GPU-CPU Buffer Access

### DISCUSSION THREAD

**Discussion by killer-of-Plato** [gfx-rs/wgpu#6051] [General]
Hi everyone,

I have a need for frequent high-performance transmission of consistently sized data from the GPU to the CPU. From my understanding, this process involves repeatedly mapping and unmapping intermediate buffers.

I am curious if there is a technical constraint that prevents simultaneous GPU and CPU access to a mapped buffer, assuming competent synchronization is in place. Is this limitation primarily a safeguard against potential issues (foot-gunning), or are there deeper technical reasons behind this restriction?

Thank you for your insights!

**Comment by teoxoy** [comment]
> Is this limitation primarily a safeguard against potential issues (foot-gunning), or are there deeper technical reasons behind this restriction?

As far as I know this is the main reason. The native APIs support persistent mapping so we could update our implementation to internally no-op unmap calls but this would have to be driven by some perf numbers showing that unmap (but excluding flush) is expensive.

**Reply by killer-of-Plato** [reply]
My assumption was not that unmap itself was costly but the continuous mapping is. I am needing to capture 60 fps textures and find it most pragmatic to unmap inter. buffer >> copy data >> remap inter. buffer, then use a dedicated thread to copy the data to cpu memory. It seems remapping every frame would be highly suboptimal, though probably technically feasible. (It would use two pairs of intermediate buffers / dedicated threads to avoid halting the render loop. I struggle to conceive of how one would do this differently.

**Reply by teoxoy** [reply]
> My assumption was not that unmap itself was costly but the continuous mapping is.

Right, I should have said map + unmap.

**Reply by killer-of-Plato** [reply]
Perhaps a separate unsafe mapping protocol would be justified?

**Reply by teoxoy** [reply]
We can still do this optimization internally. If the 2nd call to map has the same range as the previous one, do nothing, if it's different, unmap then map again. I think we still need perf numbers showing that it would be faster to land a PR doing this.

**Reply by killer-of-Plato** [reply]
I am unsure if I would be able to provide meaningful data as I assume the flush you mention is not currently accessible for comparison.

**Reply by teoxoy** [reply]
This would require implementing the internal changes and then benchmarking.

It doesn't seem terribly complicated to make the changes, I can give it a shot. That is if you'd be up for a PR adding the benchmarks first.

**Reply by cwfitzgerald** [reply]
Imo this doesn't need benchmarks, as it'll actually make the internals simpler, not more complicated

**Reply by teoxoy** [reply]
It will make internals simpler but the calls won't be as straight-forward as they are now.
